### ISSUE TITLE
"Name or service not known" error in balancer

### ISSUE THREAD

**Issue by ksa-real** [manticoresoftware/manticoresearch-helm#16]
I got a lot of error logs in the balancer. This looks pretty suspicious. Same issue in the auto-replication repo:      
https://github.com/djklim87/manticoresearch-auto-replication/issues/3      
      
As I'm not sure which one is in charge, filing to both.

**Comment by ksa-real** [comment]
After looking closely:
- In 2 deployments the balancer had this issue.
- One more deployment has 'already running` warnings:
```
2022-07-03T18:10:26-07:00 ==>	Another process observer is already running
2022-07-03T18:10:31-07:00 ==>	Another process observer is already running
2022-07-03T18:10:33-07:00 ==>	Another process **optimize** is already running
2022-07-03T18:10:36-07:00 ==>	Another process observer is already running
2022-07-03T18:10:41-07:00 ==>	Another process observer is already running
2022-07-03T18:10:46-07:00 ==>	Another process observer is already running
2022-07-03T18:10:51-07:00 ==>	Another process observer is already running
2022-07-03T18:10:56-07:00 ==>	Another process observer is already running
2022-07-03T18:11:01-07:00 ==>	Another process observer is already running
```
- The last deployment has both

I tried to investigate and put few extra logs into `optimize.php`. It was somewhat hard as there are multiple instances of `optimize` and `observer` all logging into the same place. I added logging in the beginning of both PHP files. Noticed that both files were picked up by scripts. After I killed `optimize_runner.sh` and `supervisord` brought it back, the `getaddrinfo failed` message disappeared. After I killed `php optimize.php` process, the `Another process optimize is already running` disappeared. My guess is that some conditions force Mysql connector to try to indefinitely fail and connect.

**Comment by ksa-real** [comment]
After looking at `ps aux` confirmed that both optimizer and observer got stuck in my last two deployments. I.e. the processes are running already for couple of weeks instead of finish and being re-run every couple seconds. Let me know if I can check something to help figure out what is going on.

**Comment by ksa-real** [comment]
Put a bunch of checkpoint logs into observer. Then killed just the observer. Then:
- Observer is started by job_runner.sh. This instance works up to the end and finishes.
- Next observer is stuck at `$manticoreStatefulsets = $api->getManticorePods();`, particularly `call` method.
- Next observer expectedly cannot acquire the lock.

**Comment by ksa-real** [comment]
Note, that instead of filtering out pods after retrieval it is much more efficient to filter them at API server level. For that it is necessary to add `?labelSelector=name%3D${nameLabelSelector}` to API server url.

**Comment by githubmanticore** [comment]
➤ Klim Todrik commented:
 

 
How many balancer instances do you have?

**Comment by ksa-real** [comment]
Just one balancer per release/environment.
The requested data per worker is below.
```
+------+------------------+-------+-------+-------------------------------------------------------+--------+---------------+-----------+---------------+----------------+-----------+---------------+--------------+---------+-----------------------------------------------+
| Tid  | Name             | Proto | State | Host                                                  | ConnID | Time          | Work time | Work time CPU | Thd efficiency | Jobs done | Last job took | In idle      | Chain   | Info                                          |
+------+------------------+-------+-------+-------------------------------------------------------+--------+---------------+-----------+---------------+----------------+-----------+---------------+--------------+---------+-----------------------------------------------+
|   44 | manticore_repl_0 |       | -     | wsrep 10.244.1.227:9312,10.244.1.227:9315:replication |     -1 | 375579.422164 | 6s        | 0us           |          0.00% |        25 | 4d            | No (working) | Repl    |                                               |
|   21 | work_0           |       | -     |                                                       |     -1 | 375580.467645 | 7s        | 0us           |          0.00% |    136850 | 16us          | 1s ago       |         |                                               |
|   22 | work_1           |       | -     |                                                       |     -1 | 375580.467638 | 7s        | 0us           |          0.00% |    136891 | 44us          | 1s ago       |         |                                               |
|   23 | work_2           |       | -     |                                                       |     -1 | 375580.467609 | 7s        | 0us           |          0.00% |    136961 | 16us          | 3s ago       |         |                                               |
|   24 | work_3           |       | -     |                                                       |     -1 | 375580.467589 | 7s        | 0us           |          0.00% |    136886 | 4us           | 2s ago       |         |                                               |
|   25 | work_4           |       | -     |                                                       |     -1 | 375580.467576 | 7s        | 0us           |          0.00% |    136714 | 158us         | 925ms ago    |         |                                               |
|   26 | work_5           |       | -     |                                                       |     -1 | 375580.467549 | 7s        | 0us           |          0.00% |    136906 | 22us          | 353ms ago    |         |                                               |
|   27 | work_6           |       | -     |                                                       |     -1 | 375580.467535 | 7s        | 0us           |          0.00% |    136819 | 61us          | 3s ago       |         |                                               |
|   28 | work_7           |       | -     |                                                       |     -1 | 375580.467525 | 7s        | 0us           |          0.00% |    136814 | 111us         | 2s ago       |         |                                               |
|   29 | work_8           |       | -     |                                                       |     -1 | 375580.467503 | 7s        | 0us           |          0.00% |    136820 | 20us          | 353ms ago    |         |                                               |
|   30 | work_9           |       | -     |                                                       |     -1 | 375580.467486 | 7s        | 0us           |          0.00% |    136836 | 16us          | 67ms ago     |         |                                               |
|   31 | work_10          |       | -     |                                                       |     -1 | 375580.467478 | 7s        | 0us           |          0.00% |    136829 | 6us           | 1s ago       |         |                                               |
|   32 | work_11          |       | -     |                                                       |     -1 | 375580.467457 | 7s        | 0us           |          0.00% |    136808 | 16us          | 2s ago       |         |                                               |
|   33 | work_12          | mysql | query | (local)                                               | 713488 |      0.000162 | 7s        | 0us           |          0.00% |    136710 | 210us         | No (working) | Conn    | SHOW THREADS option format=all                |
|   34 | work_13          |       | -     |                                                       |     -1 | 375580.467433 | 7s        | 0us           |          0.00% |    137006 | 52us          | 353ms ago    |         |                                               |
|   35 | work_14          |       | -     |                                                       |     -1 | 375580.467426 | 7s        | 0us           |          0.00% |    136786 | 56us          | 353ms ago    |         |                                               |
|   36 | work_15          |       | -     |                                                       |     -1 | 375580.467406 | 7s        | 0us           |          0.00% |    136804 | 4us           | 67ms ago     |         |                                               |
|   39 | TickPool_0       |       | -     |                                                       |     -1 | 375580.455504 | 34s       | 0us           |          0.00% |         0 | 14us          | 223us ago    | Listen  | tick:  2147061  works:  0  state:  in polling |
|   47 | TaskW_3          |       | -     |                                                       |     -1 | 339580.464877 | 5s        | 0us           |          0.00% |    114049 | 37us          | 67ms ago     |         |                                               |
|   45 | TaskW_2          |       | -     |                                                       |     -1 | 357580.466602 | 5s        | 0us           |          0.00% |    123106 | 41us          | 2s ago       |         |                                               |
|   38 | TaskW_1          |       | -     |                                                       |     -1 | 375580.466581 | 6s        | 0us           |          0.00% |    141266 | 40us          | 1s ago       |         |                                               |
|   37 | TaskSched        |       | -     |                                                       |     -1 | 375580.467266 | 11s       | 0us           |          0.00% |    378421 | 18us          | 67ms ago     |         |                                               |
+------+------------------+-------+-------+-------------------------------------------------------+--------+---------------+-----------+---------------+----------------+-----------+---------------+--------------+---------+-----------------------------------------------+
```
At this moment I think the issue was that apiserver was not giving back the pods results sometimes. May be because of large  output. I think it is worth changing GetManticorePods to not retrieve non-manticore pods.

**Comment by ksa-real** [comment]
Just confirmed that my suggestion helped. I've added temporary modified ApiClient.php:
```
    private function getUrl($section, $noNamespace = false)
    {
        if ($noNamespace) {
            return str_replace(['{{API-URL}}', '{{API-VERSION}}', '{{API-SECTION}}'],
                [$this->apiUrl, $this->apiSections[$section], $section], self::API_URL_SCHEME_NO_NS);
        }

        return str_replace(['{{API-URL}}', '{{API-VERSION}}', '{{NAMESPACE}}', '{{API-SECTION}}'],
                [$this->apiUrl, $this->apiSections[$section], $this->namespace, $section], self::API_URL_SCHEME . '?labelSelector=name%3Dmyrelease-manticore-worker );
    }
```
Now getManticorePods() returns just 2 pods instead of 30+ and observer.php is not locking anymore.

**Comment by ksa-real** [comment]
Also, please add timeout for K8s API call just in case.

**Comment by djklim87** [comment]
Fixed in 5.0.0.2

**Comment by sanikolaev** [comment]
> Linking the issue to https://github.com/manticoresoftware/manticoresearch-helm/issues/17

Since there's another issue for this I'm closing this one.
